ppafm/cli/generateElFF.py

```diff
@@ -162,8 +162,12 @@ def main(argv=None):
         ff_kpfm_tvs0, _ = computeElFF(v_v0_aux2, lvec, n_dim, drho_kpfm, computeVpot=args.energy, tilt=args.tilt, sigma=sigma)
 
         print("Linear E to V")
-        zpos = np.linspace(lvec[0, 2] - args.z0, lvec[3, 2] - args.z0, n_dim[0])
+        zpos = np.linspace(lvec[0, 2] - args.z0, lvec[0, 2] + lvec[3, 2] - args.z0, n_dim[0])
         for i in range(n_dim[0]):
+            if zpos[i] <= 0:
+                # z position of the KPFM tip with respect to the sample must not be zero or negative
+                # Should that happen, use periodicity in z to get zpos>0
+                zpos[i] += lvec[3, 2]
```

Comment:
I'm not familiar with the KPFM calculation, but this looks a bit odd to me. If you add one lattice vector, wouldn't it just go back to the exact same place in the periodic image? Or if it's not periodic, then it's just shifted somewhere far away. Or maybe that is exactly what you want?

Reply:
Perhaps a shorter answer: The value of `zpos` is not used as a kind of index to locate the tip. Rather, a division by this value is performed on a certain field, so the difference by the lattice vector matters.

Reply:
Okay, just making sure that it's correct.

> in case the grid origin were somewhere absurdly negative

I think it would work if you just do a modulo operation: 
```python
zpos[i] %= lvec[3, 2]
```

Reply:
Hi! Sorry for the late response.
Yes the idea of zpos together with z0 is to divide the field by the distance between the metallic parts of the tip and the sample. And it cant be zero, that's why I was adding 0.1 in some lines around there.
Bu I can't recal ading a whole z vector to solve as you guys show there... So I can't really tell what was I intending there...

Anyway I am checking the changes and all looks OK for me

Reply:
Hi @aureliojgc , the thing is that not only `zpos` should be non-zero, it should be positive, as dividing by `zpos`<0 also does not make much sense. Well, it might make sense if the tip were scanning *below* the sample instead of *above* the sample, then it would just flip the direction of the electric field caused by the bias, which would be correct. But the `ppafm` code does not allow this anyway, we expect the direction from the tip to the sample to always be a direction of decreasing *z* coordinate. So, under which circumstances could `zpos`<0 happen and why I believe adding `lvec[3,2]` (the lattice vector perpendicular to the slab, essentially) is the right way to correct it then?
Imagine a periodic system with periodicity in *z* given by slab the thickness of, say, `lvec[3,2] = 20.0`. Suppose that the user did not bother to align the position of the sample with the origin of coordinates so the *z* coordinates of the atoms of the sample happen to be somewhere in the interval `10.0 < z < 15.0`. Say `z0 = 14.0` is the reasonable choice to define `zpos` then. Suppose further the user wants to see the forces and LCPD and such in the region starting from 2 angstrom above the sample and going up to 7 angstrom above the sample, which makes if from 17.0 to 22.0 in absolute coordinates. Oops, but the slab ends at `z=20.0`. Does the user have to shift the bottom of the sample from z=10.0 to z=0 and recalculate the Hartree potential and everything in the shifted geometry, so that there would be enough space above the sample? That should not be necessary: Thanks to the periodicity in *z*, the Hartree potential and everything at `z=22.0` should be the same as it is at `z=2.0`. Except that `zpos = z - z0 = 8.0` at `z=22.0` but `zpos = -12` at `z=2.0` and the latter is of course wrong. Adding `lvec[3,2] = 20.0` to `zpos = -12` corrects it to `zpos = 8.0` and everything is fine. Do you see what I mean?
I will be of course adding the lattice vector conditionally, if `zpos` were already positive, nothing would be added.